vllm/model_executor/models/bert_embedding.py

```diff
@@ -0,0 +1,399 @@
+                                       cache_config=cache_config,
+                                       quant_config=quant_config)
+        self.intermediate = BertIntermediate(config)
+        self.output = BertOutput(config)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        kv_cache: Optional[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+    ):
+        self_attention_outputs = self.attention(
+            hidden_states,
+            kv_cache,
+            attn_metadata,
+        )
+
+        output = self.feed_forward(self_attention_outputs)
+        return output
+
+    def feed_forward(self, attention_output):
+        intermediate_output = self.intermediate(attention_output)
+        layer_output = self.output(intermediate_output, attention_output)
+        return layer_output
+
+
+class BertAttention(nn.Module):
+
+    def __init__(
+        self,
+        config: BertConfig,
+        cache_config: Optional[CacheConfig] = None,
+        quant_config: Optional[QuantizationConfig] = None,
+    ):
+        super().__init__()
+        self.self = BertSelfAttention(config=config,
+                                      cache_config=cache_config,
+                                      quant_config=quant_config)
+        self.output = BertSelfOutput(config)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        kv_cache: torch.Tensor,
+        attn_metadata: AttentionMetadata,
+    ) -> torch.Tensor:
+        self_outputs = self.self(hidden_states, kv_cache, attn_metadata)
+        attn_output = self.output(self_outputs[0], hidden_states)
+        return attn_output
+
+
+class BertSelfAttention(nn.Module):
+
+    def __init__(
+        self,
+        config: BertConfig,
+        cache_config: Optional[CacheConfig] = None,
+        quant_config: Optional[QuantizationConfig] = None,
+    ):
+        super().__init__()
+        self.hidden_size = config.hidden_size
+        tp_size = get_tensor_model_parallel_world_size()
+
+        self.total_num_heads = config.num_attention_heads
+        assert self.total_num_heads % tp_size == 0
+
+        self.num_heads = self.total_num_heads // tp_size
+        self.total_num_kv_heads = self.total_num_heads
+        self.head_dim = self.hidden_size // self.total_num_heads
+        assert self.head_dim * self.total_num_heads == self.hidden_size
+
+        self.num_kv_heads = max(1, self.total_num_kv_heads // tp_size)
+
+        self.q_size = self.num_heads * self.head_dim
+        self.kv_size = self.num_kv_heads * self.head_dim
+        self.scaling = self.head_dim**-0.5
+        self.qkv_proj = QKVParallelLinear(
+            hidden_size=self.hidden_size,
+            head_size=self.head_dim,
+            total_num_heads=self.total_num_heads,
+            total_num_kv_heads=self.total_num_kv_heads,
+            bias=True,
+            quant_config=quant_config)
+
+        self.attn = Attention(
```

Comment:
We need to use bidirectional attention here rather than causal attention

Reply:
Do you mean to implement a new type of attention in vllm? Not sure if there is a way to use the current attention implementation with different parameters.

Reply:
@robertgshaw2-neuralmagic hi, I'd like to hear your thoughts on this point. The framework of BERT model is almost completed. But the BertSelfAttention output differs from transformers. After diving into the Attention implementation, there are massive changes of Attention are needed to use bidirectional attention.
I also find some related discussions(https://github.com/vllm-project/vllm/pull/3117#issuecomment-2159049996). I think this PR is depended on #4942. What's your opinions?

Reply:
@laishzh sorry missed this note!

Yes. Pull in the `BertSelfAttention` from that PR. #4942 should land this week